ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/ChainState.hs

```diff
@@ -490,3 +548,88 @@ instance Serialise (PBftVerKeyHash c) => Serialise (PBftSigner c) where
   decode = fromPair <$> decode
     where
       fromPair (slotNo, genesisKey) = PBftSigner slotNo genesisKey
+
+{-------------------------------------------------------------------------------
+  EBB Map
+-------------------------------------------------------------------------------}
+
+-- | Append an EBB
+--
+-- Its slot will be mapped to the 'lastSignedSlot'.
+appendEBB :: forall c. (PBftCrypto c, HasCallStack)
+          => SecurityParam
+          -> WindowSize
+          -> SlotNo
+          -> PBftChainState c -> PBftChainState c
+appendEBB k n newEbbSlot cs@PBftChainState{..} =
+    assertInvariant k n $
+    Exn.assert valid $
+    cs{ebbs = ebbsInsert newEbbSlot latestNonEbbSlot ebbs}
+  where
+    latestEbbSlot    = ebbsMax ebbs
+    latestNonEbbSlot = lastSignedSlot cs
+
+    valid = At newEbbSlot > max latestEbbSlot latestNonEbbSlot
```

Comment:
I don't think it can: e.g. if `newEbbSlot = latestEbbSlot` then this function would just override the existing value in the map. `assertInvariant` runs too late to know that the same key had already existed in the map. Similar for `latestNonEbbSlot`.

Basically: this assertion constrains the order of `append` and `appendEBB` operations, which are not themselves recorded in the 'ChainState'.